Enable spi_loopback test for TI cc13xx/cc26xx

vanti opened the pull request:
This PR adds spi loopback mode support to the spi driver. In addition it fixes an issue where the clock rate used for the peripheral is incorrect - the system clock rate no longer matches the CPU clock rate that is fed to the peripheral after the change to use RTC in #19232.

This work enables the spi_loopback test to be run on the cc1352r1_launchxl.

**Update**: Loopback mode support already exists in the spi driver. Instead we modify the spi loopback test to use this support.

vanti committed:
drivers: spi: use CPU clock frequency in configuration on CC13XX/CC26XX

The SPI peripheral should be configured using the CPU clock speed and
not the system clock speed. This used to be fine because they were the
same before #19232, but now that the system clock is RTC-based (which
has a different frequency), we can no longer make that assumption.

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

vanti committed:
tests: spi_loopback: make CONFIG_SPI_ASYNC optional

Making CONFIG_SPI_ASYNC optional allows the test to be run on devices
where the feature is not yet supported in the driver.

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

vanti committed:
tests: spi_loopback: add support for cc1352r1_launchxl

Adding a configuration file for the cc1352r1_launchxl to allow the test
to run on this board.

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

galak commented on drivers/spi/spi_cc13xx_cc26xx.c:
```diff
@@ -9,14 +9,16 @@
 #include "spi_context.h"
 
+/* CPU clock speed */
+#define CPU_CLOCK_HZ            ((uint32_t)48000000)
```
We should have a DT define for this.  Look at boards/arm/sam_e70_xplained/sam_e70_xplained.dts and how it sets:

```&cpu0 {
        clock-frequency = <300000000>;
};
```

This should need something like in dts_fixup.h to use `DT_CPU_CLOCK_FREQUENCY`:

```
soc/arm/atmel_sam/same70/dts_fixup.h:#define DT_CPU_CLOCK_FREQUENCY     DT_ARM_CORTEX_M7_0_CLOCK_FREQUENCY
```

vanti replied:
Thanks @galak. I'll add this to dts.

galak reviewed (dismissed):
Use DT for CPU CLK FREQ.

tbursztyka commented on drivers/spi/Kconfig.cc13xx_cc26xx:
```diff
@@ -10,3 +10,15 @@ config SPI_CC13XX_CC26XX
+	bool "Enable loopback mode for SPI 1"
+	depends on SPI_1
+	help
+	  Set loopback mode for the TI SimpleLink CC13xx / CC26xx SPI peripheral
```
These 2 sound useless, see below in the code why. 
Also, it is very bad to have per-instance config option in Kconfig now that we are using DTS for configuring driver instances.

vanti replied:
@tbursztyka Could you clarify? I don't see anything in the code below. It'd help if you can point me to an example where per-instance config option in Kconfig was avoided. In `drivers/spi/Kconfig` I still see the `SPI_0`, `SPI_1` etc. listed as separate config options for each SPI instance.

tbursztyka replied:
oh, indeed my comment about driver change did not make it... will re-comment.

spi instances in Kconfig are legacy stuff, and we should work on removing them actually.

vanti committed:
dts: specify cpu frequency for TI CC13X2/CC26X2

Add cpu clock frequency information to DTS so that it can be retrieved
in the code.

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

vanti commented:
Added cpu clock frequency to dts based on comment from @galak.

tbursztyka commented on drivers/spi/spi_cc13xx_cc26xx.c:
```diff
@@ -119,6 +119,18 @@ static int spi_cc13xx_cc26xx_configure(struct device *dev,
+		HWREG(cfg->base + SSI_O_CR1) |= SSI_CR1_LBM;
+	}
+#endif
+
```
So I meant that piece of code. 

Looks like there is no difference between lines 125 and 130 thus why the Kconfig options seem useless to me. 
There is no need to do such base address comparison.

Also the line HWREG(cfg->base + SSI_O_CR1) |= SSI_CR1_LBM; should go after line 119. Loopback is only enabled there.

vanti replied:
Thanks @tbursztyka. Now I see what you meant. You are correct, line 119 is doing the same thing.

It looks like the issue isn't so much that there is no way to enable loopback mode, but rather the loopback test is (surprisingly) not enabling it via SPI_MODE_LOOP. I'll make that change and resubmit.

vanti committed:
tests: spi_loopback: add option for LOOP mode

Adding a Kconfig option for this test to configure the SPI in LOOP
mode. This saves physical wiring for hardware that support it.

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

vanti commented:
Removed the Kconfig options to set the loopback mode, and modified spi_loopback test to set SPI_MODE_LOOP.

tbursztyka commented on tests/drivers/spi/spi_loopback/src/spi.c:
```diff
@@ -56,15 +56,15 @@ static void to_display_format(const u8_t *src, size_t size, char *dst)
 struct spi_config spi_cfg_slow = {
 	.frequency = SLOW_FREQ,
-	.operation = SPI_OP_MODE_MASTER | SPI_MODE_CPOL |
+	.operation = SPI_OP_MODE_MASTER | SPI_MODE_CPOL | SPI_MODE_LOOP |
```
Actually no. SPI_MODE_LOOP is different than what this test is meant for. This test is meant to run on hardware with a wire short-cutting MISO-MOSI lines. So it will act as if there is an actual hw slave echoing everything being sent.
SPI_MODE_LOOP mode works slightly differently depending on the controller.

That said, it would be indeed nice to test SPI_MODE_LOOP if requested by the user.
Could you modify this commit so SPI_MODE_LOOP would be conditioned upon a local Kconfig option?
(see tests/drivers/spi/spi_loopback/Kconfig)

vanti replied:
Thanks @tbursztyka. This makes sense.
